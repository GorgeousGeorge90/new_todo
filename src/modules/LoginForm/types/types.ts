
export type UserType = {
    id:string,
    name:string,
    password:string,
}


export type UserLoginType = Pick<UserType, 'name' | 'password'>

export type LoginStateType = {
    users:UserType[],
    current: string | null,
}

export type MainFormType = {
    name: string,
    password: string,
}

export type AddUserFormType= {
    new_name: string,
    new_password: string,
}

export type FormPropsType = {
    onClick:()=> void,
}

