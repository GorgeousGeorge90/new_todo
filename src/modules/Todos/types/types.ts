
export type TodoItemType = {
    id: number,
    name?: string | null,
    title: string,
    text: string,
    complete: boolean,
    avatar?: null | string,
}

export type NewTodoProps = {
    name:string,
    title:string,
    text:string,
}

export type TodosType = {
    todos:TodoItemType[],
    status: 'idle' | 'pending' | 'fulfilled' | 'rejected',
    error:null | string,
}

export type TodosProps = {
    todos:TodoItemType[],
    title:string,
    avatar:string | null,
}

export type TaskType = Omit<TodoItemType, 'id' >
export type NewTodoType = Omit<TodoItemType, 'id' | 'avatar' | 'complete'>
export type CompleteProps = Pick<TodoItemType, 'id' | 'complete'>


