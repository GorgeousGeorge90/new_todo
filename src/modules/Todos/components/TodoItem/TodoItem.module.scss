

.todo_item_container {
  @apply flex w-fit max-w-[500px] rounded-md p-3 gap-2 items-center;

  .todo_item_checkbox {
    @apply appearance-none bg-neutral-50 h-[15px] w-[15px] flex items-center justify-center border-[1px] rounded-full
    border-violet-700 checked:after:content-['âœ“'] checked:bg-violet-700 checked:border-none
    cursor-pointer text-neutral-50;
  }

  .todo_item_header {
    @apply flex flex-col items-center gap-1;

    .todo_item_ava {
        @apply h-6 w-6;
      }
    }
  }


  .todo_item_main {
    @apply flex flex-col text-neutral-800 text-xs;

    .todo_item_title {
      @apply font-bold text-base text-neutral-700;
    }

      input {
        @apply m-2  border-none bg-neutral-100;
      }
  }



  .todo_item_right {
    @apply flex flex-col items-center text-xs;

    .todo_item_ava {
      @apply h-8 w-8;
    }
  }

  .todo_item_btn {
    @apply flex flex-col gap-1 p-1 items-center cursor-pointer;
  }