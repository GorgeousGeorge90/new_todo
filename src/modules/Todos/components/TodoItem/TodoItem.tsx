import { TodoItemType } from '../../types/types';
import styles from './TodoItem.module.scss';
import { useAppDispatch } from '../../../../store/hooks';
import { completeTodo, deleteTodo, updateTodo } from '../../store/todosSlice';
import { ChangeEvent, useState } from 'react';
import { ReactComponent as Edit } from './img/edit.svg';
import { ReactComponent as Trash } from './img/trash1.svg';



const InputStyle = {
    backgroundColor:'#a78bfa',
    border:'none',
    padding:'2px',
    color:'white',
}

const TodoItem = ({id,name,title,text,complete,avatar}:TodoItemType) => {
    const dispatch = useAppDispatch()
    const [editMode,setEditMode] = useState(false)
    const [new_text, setNewText] = useState(text)

    const handleClick = () => {
        const payload = {
            id,
            complete,
        }
        dispatch(completeTodo(payload))
    }
    const onChange = (event:ChangeEvent<HTMLInputElement>) => {
        setNewText(event.target.value)
    }
    const onBlur = () => {
        const payload = {
            id,
            new_text,
        }
        dispatch(updateTodo(payload))
        setEditMode(false)
    }

    return (<li className={styles.todo_item_container}
                style={{backgroundColor: complete ? '#a78bfa':'#ddd6fe'}}>
        <header className={styles.todo_item_header}>
            {
                avatar ?  <img className={styles.todo_item_ava}
                               src={avatar}
                               alt='avatar'/>:null
            }
            <input className={styles.todo_item_checkbox}
                   type='checkbox'
                   defaultChecked={complete}
                   onClick={handleClick}
            />
        </header>
        <main className={styles.todo_item_main}>
            <h3 className={styles.todo_item_title}
                style={{textDecoration: complete ? 'line-through':'none'}}
            >{title}</h3>
            {
                editMode ? <input type='text'
                                  value={new_text}
                                  autoFocus={true}
                                  onChange={onChange}
                                  onBlur={onBlur}
                                  style={InputStyle}
                    />: <p style={{textDecoration: complete ? 'line-through':'none'}}>{text}</p>
            }
        </main>
        <div className={styles.todo_item_btn}>
            <Edit height={'15px'}
                  width={'15px'}
                  fill={'darkslateblue'}
                  onClick={()=> complete ? console.log('disabled'):setEditMode(true)}
            />
            <Trash height={'15px'}
                   width={'15px'}
                   onClick={()=>dispatch(deleteTodo(id))}
            />
        </div>
    </li>)
}

export default TodoItem