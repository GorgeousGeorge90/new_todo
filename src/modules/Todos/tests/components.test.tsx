import * as reduxHooks from 'react-redux';
import { TodoItemType } from '../types/types';
import { render } from "@testing-library/react";
import TodoItem from "../components/TodoItem/TodoItem";


jest.mock('react-redux')
const mockedDispatch = jest.spyOn(reduxHooks, 'useDispatch')
const mockedSelector = jest.spyOn(reduxHooks, 'useSelector')

describe('render Todos',() => {
    it('goal: render TodoItem component',()=> {
        const data:TodoItemType = {
            id:1,
            title:'Hello',
            text:'nice',
            complete:false,
            avatar:null,
        }
        const dispatch = jest.fn()
        mockedDispatch.mockReturnValue(dispatch)
        mockedSelector.mockReturnValue(data)

        const view = render(<TodoItem {...data}/>)
        expect(view).toMatchSnapshot()
    })
})