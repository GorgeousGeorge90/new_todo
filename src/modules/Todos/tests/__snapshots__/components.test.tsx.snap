// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Todos goal: render TodoItem component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        class="todo_item_container"
        style="background-color: rgb(221, 214, 254);"
      >
        <header
          class="todo_item_header"
        >
          <input
            class="todo_item_checkbox"
            type="checkbox"
          />
        </header>
        <main
          class="todo_item_main"
        >
          <h3
            class="todo_item_title"
            style="text-decoration: none;"
          >
            Hello
          </h3>
          <p
            style="text-decoration: none;"
          >
            nice
          </p>
        </main>
        <div
          class="todo_item_btn"
        >
          <svg
            fill="darkslateblue"
            height="15px"
            width="15px"
          >
            edit.svg
          </svg>
          <svg
            height="15px"
            width="15px"
          >
            trash1.svg
          </svg>
        </div>
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class="todo_item_container"
      style="background-color: rgb(221, 214, 254);"
    >
      <header
        class="todo_item_header"
      >
        <input
          class="todo_item_checkbox"
          type="checkbox"
        />
      </header>
      <main
        class="todo_item_main"
      >
        <h3
          class="todo_item_title"
          style="text-decoration: none;"
        >
          Hello
        </h3>
        <p
          style="text-decoration: none;"
        >
          nice
        </p>
      </main>
      <div
        class="todo_item_btn"
      >
        <svg
          fill="darkslateblue"
          height="15px"
          width="15px"
        >
          edit.svg
        </svg>
        <svg
          height="15px"
          width="15px"
        >
          trash1.svg
        </svg>
      </div>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
